### PR TITLE
do not hide exception when ldap server has a hiccup

### LINKED ISSUE #13862: Throw exception in initMountPoints if user manager returned null
If for some reason the user manager did not find the expected user here https://github.com/owncloud/core/blob/stable7/lib/private/files/filesystem.php#L332 the logic will still go in the else block and create a bogus "local::" home entry.

This code needs to be made more robust: add a check whether `$user` is non-null and non-empty and if `$userObject` is null, which means that the user manager did not find the user, just throw an exception. This will avoid unwanted side-effects.
See https://github.com/owncloud/core/issues/12801#issuecomment-72667623 where this was found

CC @icewind1991 @blizzz

### PR DESCRIPTION
**Pull request by butonic**
this PR allows connection related exceptions in user_ldap to bubble up. When triggered by the filecache post write hook the following message will now show up in tho oc log.

```
{"reqId":"551a86144fa46","app":"hook","message":"error while running hook (\\OC\\Files\\Cache\\Shared_Updater::writeHook): No LDAP Connection to server localhost","level":3,"
```

When a user tries to login and the ldap server is not available he will now see
bildschirmfoto von 2015-03-31 13 47 16

instead of

bildschirmfoto von 2015-03-31 13 50 53

Fixes #13862

### CODE REVIEW

**Comment by LukasReschke** [comment]
Can we show the full stacktrace only if Debug mode is enabled? – It's fine to show "No LDAP Connection to server" and log this properly but we really should not expose all the other rest if not really required as I already see people complaining about exposing the directory structure etc…

**Comment by DeepDiver1975** [comment]
Looks reasonable - @butonic do we get prs against master and stable8 as well? THX

**Comment by karlitschek** [comment]
nice :+1: But I agree with @LukasReschke here. Can we make this debug only?

**Comment by butonic** [comment]
bildschirmfoto von 2015-03-31 16 10 42

**Comment by DeepDiver1975** [comment]
from my pov there is no reason to have the login page displayed - the regular exception page is quite okay - exception details should be hidden as soon as debug is disabled 

@LukasReschke or is it necessary to throw a HintException here? THX

**Comment by LukasReschke** [comment]
> @LukasReschke or is it necessary to throw a HintException here? THX

Nope. HintException's are the only ones where the actual error message is shown to the user. So a `throw new \HintException('Foo')` would show `Foo` instead of an error page to the user. This is for example used in the setup for checks like the dependencies etc.

Any other uncatched exception will print the regular exception page which shows the "Instance has encountered a problem" button as well as the request ID to allow easier debugging of the error. If debug mode is enabled the whole stacktrace will be shown to the user. (otherwise the admin needs to cross-reference the request ID with the error log.

**Comment by Xenopathic** [comment]
Could this result in the case where the LDAP server goes down (perhaps triggered by an admin), then the admin needs to go into oC to disable LDAP, but can't because login throws an exception that isn't caught?

**Comment by butonic** [comment]
There seems to be a problem with accessing the users page. @blizzz I always get a `Could not create User instance` error because the code tries to look up a user in the oc_ldap_user_mapping table. He calls `$this->access->username2dn('35946d96-6bd8-1034-90e3-ab64516f8ff1');` but that owncloud_name does not exist. when are they added to the table? Here is the full stacktrace:
```
Could not create User instance
#0 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/lib/user/manager.php(193): OCA\user_ldap\lib\user\Manager->createInstancyByUserName('35946d96-6bd8-1...')
#1 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/user_ldap.php(186): OCA\user_ldap\lib\user\Manager->get('35946d96-6bd8-1...')
#2 /home/jfd/Repositories/oc/core-stable7/lib/private/user/manager.php(104): OCA\user_ldap\USER_LDAP->userExists('35946d96-6bd8-1...')
#3 /home/jfd/Repositories/oc/core-stable7/lib/private/user/manager.php(133): OC\User\Manager->get('35946d96-6bd8-1...')
#4 /home/jfd/Repositories/oc/core-stable7/lib/private/user.php(583): OC\User\Manager->userExists('35946d96-6bd8-1...')
#5 /home/jfd/Repositories/oc/core-stable7/lib/public/user.php(92): OC_User::userExists('35946d96-6bd8-1...', NULL)
#6 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/lib/access.php(414): OCP\User::userExists('35946d96-6bd8-1...')
#7 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/lib/access.php(510): OCA\user_ldap\lib\Access->dn2ocname('uid=zombie317,o...', 'Appel, Frederic', true)
#8 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/lib/access.php(441): OCA\user_ldap\lib\Access->ldap2ownCloudNames(Array, true)
#9 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/user_ldap.php(141): OCA\user_ldap\lib\Access->ownCloudUserNames(Array)
#10 /home/jfd/Repositories/oc/core-stable7/apps/user_ldap/user_ldap.php(327): OCA\user_ldap\USER_LDAP->getUsers('', 50, NULL)
#11 /home/jfd/Repositories/oc/core-stable7/lib/private/user/manager.php(215): OCA\user_ldap\USER_LDAP->getDisplayNames('', 50, NULL)
#12 /home/jfd/Repositories/oc/core-stable7/lib/private/user.php(570): OC\User\Manager->searchDisplayName('', 50, NULL)
#13 /home/jfd/Repositories/oc/core-stable7/settings/users.php(35): OC_User::getDisplayNames('', 50)
#14 /home/jfd/Repositories/oc/core-stable7/lib/private/route/route.php(135) : runtime-created function(1): require_once('/home/jfd/Repos...')
#15 [internal function]: __lambda_func(Array)
#16 /home/jfd/Repositories/oc/core-stable7/lib/private/route/router.php(227): call_user_func('\x00lambda_2447', Array)
#17 /home/jfd/Repositories/oc/core-stable7/lib/base.php(707): OC\Route\Router->match('/settings/users')
#18 /home/jfd/Repositories/oc/core-stable7/index.php(28): OC::handleRequest()
#19 {main}
```

**Comment by blizzz** [comment]
When a new user is mapped (which happens for the first time on the users page 
or any other time a search is done and a new user is brought to ownCloud) a 
userExists is done to make sure that the desired username is not given away. 
Might this clash?

**Comment by nickvergessen** [comment]
Throwing an exception will cause strange behaviour when ajax requests are done, since they receive HTML instead of JSON? What's the best way we can/should handle this?

**Comment by LukasReschke** [comment]
> Throwing an exception will cause strange behaviour when ajax requests are done, since they receive HTML instead of JSON? What's the best way we can/should handle this?

We already have this broken behaviour in a lot of other places :-(

**Comment by MorrisJobke** [comment]
I tested this with a broken LDAP ... it works. Let me check the AJAX calls (I will test the sharing auto complete)

**Comment by butonic** [comment]
> Throwing an exception will cause strange behaviour when ajax requests are done, since they receive HTML instead of JSON? What's the best way we can/should handle this?
We could check if the accept header is application/json in index.php before printing an exception. That is actually not that much work ... but not part of this PR.

File: apps/user_ldap/lib/connection.php [hunk 1]
```diff
@@ -158,7 +158,8 @@ public function getConnectionResource() {
 			$this->establishConnection();
 		}
 		if(is_null($this->ldapConnectionRes)) {
```
Comment (MorrisJobke) [inline apps/user_ldap/lib/connection.php]: To drop (roughly) 1 out of 20 requests change that line to:
```if(rand(0,19) === 0 || is_null($this->ldapConnectionRes)) {```

File: lib/private/hook.php [hunk 2]
```diff
@@ -80,6 +80,9 @@ static public function emit( $signalclass, $signalname, $params = array()) {
 				OC_Log::write('hook',
 					'error while running hook (' . $i["class"] . '::' . $i["name"] . '): '.$e->getMessage(),
 					OC_Log::ERROR);
+				if($e instanceof \OC\ServerNotAvailableException && $signalclass === 'OC_Filesystem' && $signalname === 'setup') {
+					throw $e;
+				}
```
Comment (MorrisJobke) [inline lib/private/hook.php]: @icewind1991 Can you have a look at this?

**Comment by MorrisJobke** [comment]
How to test this:

1. Setup a LDAP config with users
2. Login as an LDAP user and share a file with the admin (non-LDAP user)
3. Shut down the LDAP server
4. Do a PROPFIND on the root folder (with the share from a LDAP user)

#### current behaviour:

The shared folder is simply omited.

With running LDAP server:
```
$ curl -X PROPFIND -H "Depth: 1" -u mjob:mjob http://localhost/master/remote.php/webdav/  | xmllint --format -
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/master/remote.php/webdav/</d:href>
    ...
  </d:response>
  <d:response>
    <d:href>/master/remote.php/webdav/Shared-from-LDAP-user/</d:href>
    ...
  </d:response>
</d:multistatus>
```

With stopped LDAP server:
```
$ curl -X PROPFIND -H "Depth: 1" -u mjob:mjob http://localhost/master/remote.php/webdav/  | xmllint --format -
<?xml version="1.0" encoding="utf-8"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:oc="http://owncloud.org/ns">
  <d:response>
    <d:href>/master/remote.php/webdav/</d:href>
    ...
  </d:response>
</d:multistatus>
```


#### Behaviour with this patch

The PROPFIND result in a 500

```
$ curl -X PROPFIND -H "Depth: 1" -u mjob:mjob http://localhost/master/remote.php/webdav/ -v                                                                   
...
< HTTP/1.1 500 Internal Server Error
< Date: Wed, 01 Apr 2015 15:50:49 GMT
...
```

**Comment by jcfischer** [comment]
@MorrisJobke unfortunately - still creates the f***** local:// entries.

I used the `rand(0,100) === 0` method to simulate pathological conditions and let it run on our internal sync server. (only a few users running their sync clients this time of day). Within 20 minutes, 2 new local:// entries (probably due to my sync client doing discovery)

**Comment by jcfischer** [comment]
I assume there is another code path that eats the exception. I have now disabled the creation of the local:// entry and will try to capture a backtrace of it happening

**Comment by MorrisJobke** [comment]
The idea now is that if no user could be fetched from any backend we simply cannot mount anything. We should therefore not mount a local storage based on the userId. This simply is wrong. @icewind1991 Can you confirm that this is how it should work? https://github.com/owncloud/core/commit/4a87986362370ec85cec83192fc578c692d01b0c?w=1

We throw an exception and let the endpoint (webdav, web UI) know that something bad happened. We later should try to make the backends more robust to failures on the underlying infrastructure, but for now we not have a way to communicate this up to the mount point init.

**Comment by butonic** [comment]
@owncloud-bot Retest this, please.

**Comment by MorrisJobke** [comment]
Jenkins PR: #15376

**Comment by karlitschek** [comment]
I agree. It LDAP doesn´t respond with a proper user then it doens´t make sense to try to do workarounds.  The server should just log the error, return an error and exit.

**Comment by MorrisJobke** [comment]
Jenkins is happy https://github.com/owncloud/core/pull/15376#issuecomment-88885006


Refer to this link for build results (access rights to CI server needed): 
https://ci.owncloud.org//job/pull-request-analyser-ng-simple/11195/
:rocket: Test PASSed.:rocket:
chuck

**Comment by MorrisJobke** [comment]
Fixes #13862

**Comment by MorrisJobke** [comment]
Still a happy Jenkins https://github.com/owncloud/core/pull/15376#issuecomment-88908872


Refer to this link for build results (access rights to CI server needed): 
https://ci.owncloud.org//job/pull-request-analyser-ng-simple/11199/
:rocket: Test PASSed.:rocket:
chuck

**Comment by MorrisJobke** [comment]
@PVince81 @blizzz @nickvergessen Can we get this reviewed?

File: lib/private/user/nouserexception.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+ * This file is licensed under the Affero General Public License version 3 or
+ * later. See the COPYING-AGPL file.
+ *
+ * @author Jörn Friedrich Dreyer <jfd@owncloud.com>
+ * @copyright Jörn Friedrich Dreyer 2015
+ */
+
+namespace OC\User;
+
+class NoUserException extends \Exception {}
```
Comment (nickvergessen) [inline lib/private/user/nouserexception.php]: shouldn't this be in the public namespace, so apps can deal with it, without conflicting the the app code checker? :see_no_evil:

File: tests/lib/files/filesystem.php [hunk 4]
```diff
@@ -22,6 +22,8 @@
 
 namespace Test\Files;
 
+use OC\User\NoUserException;
```
Comment (nickvergessen) [inline tests/lib/files/filesystem.php]: not required

File: lib/private/user/nouserexception.php (hunk 3 above)
Comment (nickvergessen) [inline lib/private/user/nouserexception.php]: Ah I see this applies to all the exceptions, so we can deal with that in another issue/pr

**Comment by MorrisJobke** [comment]
Rebased. Can we get this in. There are more and more problems with stable7 instances that could be solved by this.

cc @nickvergessen @LukasReschke @blizzz @icewind1991

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by karlitschek** [comment]
:+1:

**Comment by blizzz** [comment]
If this was resolved https://github.com/owncloud/core/pull/15332#issuecomment-88200947

Checking

**Comment by blizzz** [comment]
The exception is not thrown, when the connection was established initially, but got lost during operation.

For example, do a search using occ over many users and shut off LDAP inbetween. Only partly results will be returned, »LDAP error Can't contact LDAP server (-1) after calling ldap_read« is being logged. I'll have a look.

**Comment by blizzz** [comment]
added throwing another one. @butonic @MorrisJobke  you ok with this?

**Comment by butonic** [comment]
@blizzz if you also port it to oc8 & master I am ok with it

**Comment by blizzz** [comment]
will do

**Comment by blizzz** [comment]
Added to the stable8 PR and, since master is done, opened an additional PR therefore: https://github.com/owncloud/core/pull/15811

**Comment by MorrisJobke** [comment]
:+1:

**Pull request merged**